### PR TITLE
[WIP] Implement BanData

### PR DESCRIPTION
**Pull request by jamierocks**
As this is my first time implementing something from the Data API, I am making the process open so I don't make a mess of things.

- [x] SpongeBanData - complete
- [x] ImmutableSpongeBanData - complete
- [x] BanDataProccesor - should be complete
- [x] UserBansValueProccesor - should be complete
- [x] Register everything - correct me if i have missed something

Upon completion I will squash all the commits, unless you'd rather do that.

tldr; please correct me. pr is not yet finished.

### CODE REVIEW

File: src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeBanData.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+import org.spongepowered.api.data.value.immutable.ImmutableSetValue;
+import org.spongepowered.api.data.value.immutable.ImmutableValue;
+import org.spongepowered.api.util.ban.Ban;
+import org.spongepowered.common.data.manipulator.immutable.common.AbstractImmutableSingleData;
+import org.spongepowered.common.data.manipulator.mutable.entity.SpongeBanData;
+import org.spongepowered.common.data.value.immutable.ImmutableSpongeSetValue;
+
+import java.util.Set;
+
+public class ImmutableSpongeBanData extends AbstractImmutableSingleData<Set, ImmutableBanData, BanData>
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeBanData.java]: This isn't just a `Set`, it's a `Set<Ban>`.

File: src/main/java/org/spongepowered/common/data/manipulator/mutable/entity/SpongeBanData.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+import org.spongepowered.api.util.ban.Ban;
+import org.spongepowered.common.data.manipulator.immutable.entity.ImmutableSpongeBanData;
+import org.spongepowered.common.data.manipulator.mutable.common.AbstractSingleData;
+import org.spongepowered.common.data.value.mutable.SpongeSetValue;
+
+import java.util.Set;
+
+public class SpongeBanData extends AbstractSingleData<Set, BanData, ImmutableBanData> implements BanData {
+
+    private final Set<Ban.User> bans;
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/manipulator/mutable/entity/SpongeBanData.java]: You don't need to have this. `AbstractSingleData` has this in a field for you.

File: src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeBanData.java (hunk 1 above)
Comment (jamierocks) [inline src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeBanData.java]: I'm pretty sure thats what I did :S 
Fixing now.

File: src/main/java/org/spongepowered/common/data/manipulator/mutable/entity/SpongeBanData.java (hunk 2 above)
Comment (jamierocks) [inline src/main/java/org/spongepowered/common/data/manipulator/mutable/entity/SpongeBanData.java]: Oh cool.

**Comment by jamierocks** [comment]
This should be ready to merge, would you like me to squash the commits?

File: src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeBanData.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    @Override
+    public ImmutableSetValue<Ban.User> bans() {
+        return new ImmutableSpongeSetValue<Ban.User>(Keys.USER_BANS, this.getValue());
+    }
+
+    @Override
+    public int compareTo(ImmutableBanData o) {
+        return 0;
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeBanData.java]: no comparison?

File: src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeBanData.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+import org.spongepowered.api.data.value.immutable.ImmutableSetValue;
+import org.spongepowered.api.data.value.immutable.ImmutableValue;
+import org.spongepowered.api.util.ban.Ban;
+import org.spongepowered.common.data.manipulator.immutable.common.AbstractImmutableSingleData;
+import org.spongepowered.common.data.manipulator.mutable.entity.SpongeBanData;
+import org.spongepowered.common.data.value.immutable.ImmutableSpongeSetValue;
+
+import java.util.Set;
+
+public class ImmutableSpongeBanData extends AbstractImmutableSingleData<Set<Ban.User>, ImmutableBanData, BanData>
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeBanData.java]: To be honest, you shouldn't actually extend SingleData because you want to make defensive copies of the `Set` and use `ImmutableSet`. The same applies to `SpongeBanData`.

File: src/main/java/org/spongepowered/common/data/processor/value/entity/UserBansValueProcessor.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+    @Override
+    public boolean supports(ValueContainer<?> container) {
+        return container instanceof User;
+    }
+
+    @Override
+    public DataTransactionResult transform(ValueContainer<?> container,
+            Function<Set<Ban.User>, Set<Ban.User>> function) {
+        return null;
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/value/entity/UserBansValueProcessor.java]: This isn't finished.

File: src/main/java/org/spongepowered/common/data/processor/value/entity/UserBansValueProcessor.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+    @Override
+    public DataTransactionResult offerToStore(ValueContainer<?> container, BaseValue<?> value) {
+        return null;
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/value/entity/UserBansValueProcessor.java]: This isn't finished.

File: src/main/java/org/spongepowered/common/data/processor/value/entity/UserBansValueProcessor.java [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+    @Override
+    public DataTransactionResult offerToStore(ValueContainer<?> container, Set<Ban.User> value) {
+        return null;
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/value/entity/UserBansValueProcessor.java]: This isn't finished.

File: src/main/java/org/spongepowered/common/data/processor/value/entity/UserBansValueProcessor.java [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+    @Override
+    public DataTransactionResult removeFrom(ValueContainer<?> container) {
+        return null;
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/value/entity/UserBansValueProcessor.java]: This isn't finished.

File: src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeBanData.java (hunk 3 above)
Comment (jamierocks) [inline src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeBanData.java]: Will do.

File: src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeBanData.java (hunk 4 above)
Comment (jamierocks) [inline src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeBanData.java]: How about just do `super(ImmutableBanData.class, Sets.newConcurrentHashSet(bans), Keys.USER_BANS);` ?

**Comment by gabizou** [comment]
Please update according to the changes recently merged.

**Comment by gabizou** [comment]
@jamierocks Am I correct in thinking that you spoke with @Aaron1011 about re-representing bans and whitelist as a service perspective instead of a Data perspective in the API?

**Comment by jamierocks** [comment]
Yes you are. As to why I haven't updated this PR.

**Comment by gabizou** [comment]
@jamierocks So I would think that this PR should be closed if such an API change is going to be made (since I did speak to him about approving the change).

**Comment by jamierocks** [comment]
Yeah, it should. (I was waiting for confirmation from you).
